### DISCUSSION TITLE
Best route to textSearch including substrings

### DISCUSSION THREAD

**Discussion by fergusmeiklejohn** [supabase/supabase#3542] [Questions] [answered]
How do we do a textSearch where user inputs "supa" and we find "Supabase" in the target column?
At the moment the textSearch will find no matches on searching for "supa", while users used to google searches would expect substring search to find the full word.
I found this Trigram module that looks like it might work for this use case: https://www.postgresql.org/docs/current/pgtrgm.html

What do experts think about this? What should be our strategy to do full text search including substrings?

**Comment by dthyresson** [comment, marked as answer]
Hi @fergusmeiklejohn Full text search on Postgres relies on lexemes -- little parts of words (English words) that in its smallest form can identify that word and related ones: plurals, adjective forms, cases, etc.

You can try this SQL to experiment with lexemes:

```sql
-- acheive lexeme

SELECT
  to_tsvector(w.word)
FROM (
  SELECT
    'achievement' AS word) AS w;


SELECT
  *
FROM (
  SELECT
    'achievement' AS word) AS w
WHERE
  to_tsvector(w.word) @@ to_tsquery('achieve');

SELECT
  *
FROM (
  SELECT
    'Achievements' AS word) AS w
WHERE
  to_tsvector(w.word) @@ to_tsquery('achieve');

SELECT
  *
FROM (
  SELECT
    'Achiever' AS word) AS w
WHERE
  to_tsvector(w.word) @@ to_tsquery('Achieve');

-- supabase

SELECT
  to_tsvector(w.word)
FROM (
  SELECT
    'Supabase' AS word) AS w;


SELECT
  *
FROM (
  SELECT
    'Supabase' AS word) AS w
WHERE
  to_tsvector(w.word) @@ to_tsquery('supabases');
```

You can see that "achievement" is `'achiev':1` and this can match `achievement`, `Achiever` `achievements` in a search query.

Supabase however is not a general English word, so its lexeme is `supabas:1` so it will also match a plural 'Supabases'.

You might want to explore fuzzy (enable the `FUZZYSTRMATCH` extension) search approaches instead of full text -- and also the trigram.

See: https://www.postgresql.org/docs/13/fuzzystrmatch.html

```sql
SELECT soundex('Supabase'), soundex('Supa'), difference('Supabase', 'Supa');
```

And if the difference in the sounded is within a threshold, it is a match.

Or

```sql
SELECT levenshtein('Supabase', 'Supa');
```

If you enable `PG_TRGM` you can as you note do:

```sql
SELECT word_similarity('Supabase', 'Supa');
```

**Reply by fergusmeiklejohn** [reply]
This is a brilliant answer! Thanks very much :-)
